aten/src/ATen/ParallelThreadPoolNative.cpp

```diff
@@ -67,10 +68,16 @@ int get_num_interop_threads() {
 }
 
 void launch(std::function<void()> func) {
+  auto debug_info = getThreadLocalDebugInfo();
+  auto fn = [func, debug_info]() {
+    auto prev_info = setThreadLocalDebugInfo(std::move(debug_info));
+    func();
+    setThreadLocalDebugInfo(std::move(prev_info));
```

Comment:
This doesn't look exception safe either.

Reply:
is it the shared_ptr move the concern here?

Reply:
I guess now it's fine with noexcept. I'm curious about general exception handling in at::parallel. Do we rely on `func` to never throw? (otherwise it'd lead to hard-to-debug bugs). In any case - would be nice to add a comment about exception handling (it's a common pitfall - I can easily recall 3-4 times when we messed it up)

Reply:
@dzhulgakov, we have three implementations of parallel_for (OpenMP, TBB and native one) and all of them capture exceptions from the given functor (from within the subtasks) and then rethrow the exception outside of parallel_for

Reply:
I was referring to the `func()` call. If the invariant on `launch` is that the passed in functor should never throw, that should be documented on the function.